Add Nixpkgs invocation helper modules

roberth opened the pull request:
###### Description of changes

This makes a number of well tested and documented additions that lead up to

 - A reusable module that module system applications can use for the purpose of invoking the Nixpkgs function.
 - A much sought after flake-parts module.

... using the supporting additions and improvements:

 - `lib.systems.equals` function for comparing elaborated system strings.
 - `lib.systems.toLosslessStringMaybe` which is useful for converting elaborated systems to _optional_ attribute names.
 - The `lib.systems` tests are now run by ofborg.

This PR contains commits that have been split into a separate PR
 - #237512

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
pkgs/top-level/module: init

roberth committed:
flake.nix: Add modules.generic.nixpkgs

A helper for module system applications.

Naming compatible with https://github.com/NixOS/nix/issues/6257
To be supported by nix flake check in https://github.com/NixOS/nix/pull/8332
Concept is in the spirit of a proposed module based solution to [RFC 0078 System-agnostic configuration file generators](https://github.com/NixOS/rfcs/pull/78).

roberth committed:
modules.generic.nixpkgs: Add optimization machinery

roberth committed:
flake.nix: Optimize modules.generic.nixpkgs when pkgs is legacyPackages

roberth committed:
nixos/doc: Add modules.generic.nixpkgs to release notes

fricklerhandwerk commented on doc/using/configuration.chapter.md:
```diff
@@ -189,6 +189,13 @@ The following attributes can be passed in [`config`](#chap-packageconfig).
 
+## Calling Nixpkgs from the Module System {#sec-invoke-nixpkgs-by-module}
+
+For [module system](#module-system) application authors, Nixpkgs provides a reusable module for the purpose of invoking Nixpkgs. Most users should not have to be aware of this. The documentation for its options is meant to be re-exposed in the application's documentation and is therefore not repeated here.
```
By "module system application" you mean things like Home Manager? I'm not sure what you mean here.

roberth replied:
That's what I meant.
I don't know how to say this without going on a tangent. I think going on a tangent to explain this would distract from the content. Most users will know to skip the clause or skip the paragraph.

fricklerhandwerk replied:
```suggestion
For authors of applications based on the [module system](#module-system), Nixpkgs provides a reusable module for the purpose of invoking Nixpkgs. Most users should not have to be aware of this. The documentation for its options is meant to be re-exposed in the application's documentation and is therefore not repeated here.
```

fricklerhandwerk commented on doc/using/configuration.chapter.md:
```diff
@@ -189,6 +189,13 @@ The following attributes can be passed in [`config`](#chap-packageconfig).
+
+For [module system](#module-system) application authors, Nixpkgs provides a reusable module for the purpose of invoking Nixpkgs. Most users should not have to be aware of this. The documentation for its options is meant to be re-exposed in the application's documentation and is therefore not repeated here.
+
+The flake attribute for this module is `.modules.generic.nixpkgs`.
+
+Non-flake applications may import the implementation file [`"${nixpkgs}/pkgs/top-level/module/module.nix"`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/module/module.nix).
```
```suggestion
Applications can import the implementation file [`"${nixpkgs}/pkgs/top-level/module/module.nix"`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/module/module.nix).

The flake attribute (experimental) for this module is `.modules.generic.nixpkgs`.
```

I would oppose introducing the same "non-X" terminology as has proliferated for NixOS (calling everything outside NixOS a "non-NixOS" system as if NixOS was the only thing that existed). Also flakes should not be documented as if they were stable and the first thing to consider even thinking about.

fricklerhandwerk commented on pkgs/top-level/module/module.nix:
```diff
@@ -0,0 +1,176 @@
+        Specifies the platform where the package will be built. This is
+        optional; by default Nixpkgs will perform native compilation.
+
+        By default, packages are built on the system where it runs, but
+        you can change where they're built. Setting this option will
+        cause NixOS to be cross-compiled.
```
```suggestion
        Setting this option will cause NixOS to be cross-compiled.
```

The first sentence is redundant with the one before.

fricklerhandwerk commented on pkgs/top-level/module/module.nix:
```diff
@@ -0,0 +1,176 @@
+        '';
+      type = types.listOf overlayType;
+      description = lib.mdDoc ''
+        List of overlays to use with the Nix Packages collection.
```
```suggestion
        List of overlays to use with Nixpkgs.
```

fricklerhandwerk reviewed (commented):
Just a superficial review. TBH the documentation doesn't explain to me who and what this thing is for exactly, and how it's supposed to be used. I could go into option documentation if you like, once the big picture is clear.

Ericson2314 commented on pkgs/top-level/module/tests.nix:
```diff
@@ -0,0 +1,115 @@
+# A the tests are invoked using an already invoked nixpkgs, we're going to
+# mostly ignore this pkgs.
+# To catch silly mistakes, let's rename this to pkgs_.
+{ pkgs }:
+let pkgs_ = pkgs;
+in
+let pkgs = throw "You probably didn't mean to use pkgs, but rather use a more specific pkgs variable in this test.";
+
+    inherit (pkgs_) lib emptyFile;
+    inherit (lib) evalModules;
+in
```
Now that the code is written, we know what is needed :)
```suggestion
{ lib, emptyFile }:
```

Ericson2314 commented on pkgs/top-level/module/tests.nix:
```diff
@@ -0,0 +1,115 @@
+        # All options that do not directly affect the construction of pkgs.
+        # If a new option's effect is implemented by setting one of the options
+        # already in `inputsSet`, it does not need to be listed here.
+        "pkgs" "_optimize" "_module"
```
```suggestion

        # Defined by us
        "pkgs" "_optimize"
        
        # From the module system itself
        "_module"
```

Ericson2314 commented on pkgs/top-level/module/tests.nix:
```diff
@@ -0,0 +1,115 @@
+        pkgs // { testData = args; };
+    };
+  };
+  allInputsSet = {
```
```suggestion

  # All inputs that effect the construction of the package set / arguments to _optimize.
  # Not `_optimize` because it is not an argument to itself.
  # Not `pkgs` because it is not an argument and the result of the `_optimize` call.
  allInputsSet = {
```

Ericson2314 reviewed (commented):
I didn't re-review the docs; I trust @fricklerhandwerk on that.

I reviewed the rest, and it looks good modulo these small few things. Excited that we are getting started on the idea behind https://github.com/NixOS/rfcs/pull/78 --- the grand module deduplication!!

roberth committed:
nixosModules.noLegacyPkgs: init

roberth committed:
nixos/qemu-vm.nix: Use types.pkgs directly

This hack isn't needed anymore, and would be incompatible with having submodule at `nixpkgs.*`.

roberth committed:
nixos/doc: Add Optional Modules

roberth commented on flake.nix:
```diff
@@ -70,6 +84,66 @@
+        };
+      };
+
+      /*
+        https://flake.parts
+        Maintainer: @roberth
+      */
+      flakeModules = {
+        pkgs = { flake-parts-lib, ... }: {
```
> * This can be turned into N+M by creating stable standards and using them instead.

flake-parts is such an interface.

> * flake-parts sould either be moved to the NixOS GitHub organization, or

I don't think that would help the project. Maintaining it is 95% maintaining a clean interface (as you alluded to), pushing back on things that don't belong in it. I wouldn't trust someone else to do that.
If for whatever reason the NixOS organization needs to take ownership of this project, it can be forked. Until then I would recommend them to delegate this work to me.

> * code can't easily be tested and documented without also involving flake-parts

I run CI for it asynchronously, twice a day and I'm happy to fix things as they come up. There's very little to integrate, so problems will be rare, based on my experience with other integrations. It's a very thin layer on top of the generic module after all.

Would you like to review the rest of the PR? I think it's been decently covered by John and Valentin, but the part about optional NixOS modules in the manual is still new, as is the `noLegacyPkgs` module.

roberth replied:
I've removed the flake-parts module for now. Flake-parts users will still be able to use the generic pkgs module.

roberth committed:
nixos/doc/manual: Set feature flag and explain minimalModules use

A feature flag warning was rightfully picked up by ofborg. This should
solve it, and hopefully help out the next person who tries to render
docs for an optional module.

blaggacao commented on doc/using/configuration.chapter.md:
```diff
@@ -191,6 +191,13 @@ list-id: configuration-variable-list
 
+## Calling Nixpkgs from the Module System {#sec-invoke-nixpkgs-by-module}
+
+For [module system](#module-system) application authors, Nixpkgs provides a reusable module for the purpose of invoking Nixpkgs. Most users should not have to be aware of this. The documentation for its options is meant to be re-exposed in the application's documentation and is therefore not repeated here.
```
```suggestion
For [module system](#module-system) application authors, Nixpkgs provides a reusable module for the purpose of configuring Nixpkgs. Other users should not have to be aware of this. Therefore, the documentation of its options is not rendered here. Module system authors should re-expose it in their own documentation.
```

blaggacao replied:
The main clarification is `configuring` instead of `invoking`. The rest just tries to amend a bit more to the reader's ductus.

roberth replied:
"Module system application authors" is precisely as intended.
I'd characterize the audience in roughly descending guesstimated size

- module system users
- module authors
- module system application authors
- module system contributors (although larger if you consider `types.nix` beyond `submodule`, which we technically should)

blaggacao replied:
If useful, for later: [diagram](https://play.d2lang.com/?script=jFI7isMwEO11ineBvYCLBbO1qyUHkO1JENgzQiOFhOC7B_lDLEihTozeG72PWu8nN9gcWYOXAToZ00S4KIVtAKwKG3gKKmyA5QP7E47B9SlKDbpNObpqIH5-SznEg4ykoIfPPpRwlfBVzIl6PDoct6Tr2XGSpHCzD3KnmTiW7k8rCpe3YMdk8xbHUcxi_p8aaS7S20Y4hVvvfed2-Utbx1U17JyqNg7B2V_RPrHtJ1LzDgAA__8%3D&)

roberth replied:
> * module system contributors (although larger if you consider `types.nix` beyond `submodule`, which we technically should)

This one is irrelevant. It's rare for the module system itself to change.

As for your diagram, it's not clear to me whether the System box refers to the module system or part of a module system application.

It's unfortunate the "system" is already part of the "module system" name. I think a system of modules would be a nice description of a module system application...

Perhaps a more interesting classification of domains is
 - **core modules**: set of modules that everything in a module system application depends on. Examples in NixOS:
    - everything in top-level
    - systemd perhaps
 - **integration modules**: modules that simplify the integration of a large number of things
    - services
    - programs
    - image formats perhaps (not sure about the status of these tbh. loading these as modules isn't always perfect)
 - **user modules**
    - a "user"'s configuration
    - any modules they wrote for their configuration
 - **module system**: mostly irrelevant; it's the air we breathe
    - `_module` options
    - a subset of `lib`

blaggacao replied:
New [diagram](https://play.d2lang.com/?script=jFPBatxADL37Kx65tD0kue8hEPbUQ6EQ-gGztmwLbMloNGmWkn8vM7Z3bTaE3MZjPT299zTP0zRwHbJlB_yrgAV0QDxHpxHaYtQmDRQr4Fc54U8km6uBMv4BE1lUqYD3a9lRxY1PyfUr1c8p-_rlQtw_7cchqbWhCHqbsshIaNU-HGYDXUnr9S_FcmZJmiJ4nExfaSTxvfpNi53KzkKTQu7C4lpVwFGNltKYUT_FqbNi-Pb6mKLruL35bdrycn5J7CWAbbfVqmti5Ju44H1w5MTP3rN0YEFYfq7hhmv6aGgiaSIudn8w6Q3ljivyOA3cnuE9Zf0XtLYIGPKuQtJ4Iss3Zai4cO3l39AE3KVIdvct7hm9pzP-mnoJO3-y5fxa7pIFv0hZvfys8RXGF9zs-w1qmtshdJ1RN2tkKbLz8tCbz5JrldrICSSvbCp5j_CdHroHjKHuWQiPqAdNDR5BXv8otO_Vslkvc0j3T9i9UpJwys-x-h8AAP__&) :smile: :handshake:

roberth replied:
Lovely. What's a suite though?

The addition of profiles is nice. User modules tend to be like profiles, but I don't know if everyone agrees on what a profile is... And we should get rid of the term because of `nix profile`.

Maybe _suite_ is a nice word for that, so I'm very curious. If you could briefly define profile and suite, that would be sweet.

blaggacao replied:
I think profile is a `haumea.load`-able. Essentially an almost-`json`-serializable. User Data. Definitions, not declarations.

For bigger projects "suites" may be handy as "profile"-aggregators. For example: the teacher suite may have the student suite + some management profiles.

roberth replied:
Ohh it has tooltips! :facepalm: 

Too bad _profile_ is already rather overloaded in the Nix world, so I'd be inclined to call them suites even if they only contribute to one option value.

blaggacao commented:
In the outer rings of context to this PR, I have attempted abstractions in a similar problem domain spanning [beyond nixpkgs here](https://github.com/divnix/hive/blob/ab85a68eea36e79de4ad0bc53919a9c7e5e98678/src/bee-module.nix#L26-L92).

Note that this comment has no bearing to this PR beyond situational awareness for interested parties.

roberth committed:
doc/using/configuration: Update

Co-authored-by: David Arnold <dgx.arnold@gmail.com>

infinisil commented on pkgs/top-level/module/module.nix:
```diff
@@ -0,0 +1,174 @@
+
+  optDefaultPrio = (lib.mkOptionDefault null).priority;
+
+  isCross = config.buildPlatform != config.hostPlatform;
+  systemArgs =
```
Let's have a comment here mentioning that this is a bit of a hack and should be moved into the pkgs entrypoint function definition

infinisil commented on pkgs/top-level/module/module.nix:
```diff
@@ -0,0 +1,174 @@
+  # key are usable with _memoize implementations that are not aware of them,
+  # as long as these additions are unused.
+  inputsSet = lib.filterAttrs
+    (_: v: v != v.isDefined && v.highestPrio < optDefaultPrio)
```
This condition needs a bit of work, `v != v.isDefined` seems wrong (detected by @roberth himself!)

infinisil commented on flake.nix:
```diff
@@ -12,6 +12,20 @@
+      pkgsMemoizer = { pkgs, inputsSet, ... }:
+        let system = lib.systems.toLosslessStringMaybe inputsSet.hostPlatform.value.system;
+        in
+        if lib.attrNames inputsSet == [ "hostPlatform" ] && system != null
```
See also the other comment

infinisil commented on flake.nix:
```diff
@@ -12,6 +12,20 @@
       forAllSystems = lib.genAttrs lib.systems.flakeExposed;
+
+      pkgsMemoizer = { pkgs, inputsSet, ... }:
+        let system = lib.systems.toLosslessStringMaybe inputsSet.hostPlatform.value.system;
```
```suggestion
        let system = lib.systems.toLosslessStringMaybe inputsSet.hostPlatform.value.system;
```

Calling this on `system` sounds wrong :laughing:

infinisil reviewed (changes requested):
Reviewed together with @roberth on a call
